### ISSUE TITLE
Can't get jupyter-vim to work with Anaconda on Windows

### ISSUE THREAD

**Issue by willamgld** [jupyter-vim/jupyter-vim#8]
Hello, I am stuck at the last step of the installation procedure, with gVim throwing an error message after calling :JupyterConnect: 
![image](https://user-images.githubusercontent.com/53537478/65057673-f26ee900-d940-11e9-961f-441b29347203.png)

I am running the latest gVim 8.1.2044 with +python/dyn and +python3/dyn, Vundle , Anaconda 3 with Python 3.7.2 and Windows 10. (Note: despite above call stack, I am _not_ using cygwin) 
I think Python is linked properly:
```
set pythonthreedll=python37.dll
set pythonthreehome=C:\ProgramData\Anaconda3
```

:python3 import sys; print(sys.version) returns 

![image](https://user-images.githubusercontent.com/53537478/65058016-8ccf2c80-d941-11e9-8f30-edc23ac4d44c.png)
I generated and edited .jupyter/*config files, as prescribed. 
I am not sure whether the following is the reason, but I am unclear as to how to properly start a Jupyter console and connect it to gVim. The instructions on GitHub for this step are for Linux, spawning a console and a vim instance from the same shell, and I am unclear how to achieve the same effect _on Windows_. To start a Jupyter console from Anaconda's prompt ($ jupyter console), I have to be in Admin mode ('access denied' when not), but then gVim started from the same console loads with (non-existent) Admin dot-files and configuration. Can you please clarify how to complete the installation?

**Comment by wmvanvliet** [comment]
What happens when you try to `import zmq` inside a regular anaconda python console?

**Comment by willamgld** [comment]
It gets imported, quietly:

![image](https://user-images.githubusercontent.com/53537478/65193114-f6396300-da47-11e9-9fb9-67448d2c0af9.png)

**Comment by wmvanvliet** [comment]
That looks fine then. Could it be a permissions problem? Why do you need to start the anaconda prompt in admin mode? Perhaps your Vim (in normal user mode, as it should be running) does not have access to the dll because there are somehow admin-only? Reinstalling anaconda in a place where your normal user can access it may help.

**Comment by willamgld** [comment]
I was eventually able to get things to work. I think the mistake was not starting gVim from the same console as the jupyter. This is pretty natural on Linux, but less obvious on Windows, with its much more convenient ways of starting gVim using file type associations or Explorer's context menu .

If this was the only problem, I would stop my post here, but there were other things in my configuration that needed to be fixed, as well. 

The reason for having to start jupyter in admin mode was, indeed, permissions-related. I had Anaconda installed for all users under C:\ProgramData\Anaconda3, and ended up giving my local user recursive full control rights (r-click on the folder->Properties->Security->Edit). Importantly, the permissions did not propagate to ...\Anaconda3\Scripts\\[your_environment]\jupyter-console.exe and jupyter.exe, so I had to grant myself full control to those files, separately. Once this was done, I could start the jupyter without elevating the console.

For the benefit of anyone having similar issues on Windows, below are my installation notes for my specific configuration of Windows 10 / Anaconda / Python 3 / gVim 8 / Vundle, as well as clarifications of some fine points I found non-intuitive, as a relative beginner. You may have to modify some steps to accommodate your configuration.

- Get a recent gVim (>8.1), perhaps a nightly build, depending on when you are reading this. There was some issue early on in the process that can't recall now, that was resolved by updating gVim. 
-  Configure .vimrc: 
-- in Vundle section of .vimrc, add:
  `Plugin 'jupyter-vim/jupyter-vim'`
--     specify location of dynamic Python:
  `" Windows`
  `set pythonthreedll=python37.dll`
  `set pythonthreehome=C:\ProgramData\Anaconda3`
-- assign `<localleader>` to an available key:
`au BufNewFile,BufRead *.py let maplocalleader = ","`
Some common alternatives include "\\\\" and "_"
**Important**: This mapping should precede the Vundle plugin import section in .vimrc; otherwise, it will have no effect on jupyter-vim shortcuts (see `:h leader` in Vim)
-- Open a new gVim session, execute 
  `:PluginInstall`

- Configure Jupyter console client to display output from other clients, as instructed by the official readme.
-- One comment here is that the auto-generated jupyter_console_config.py says that remote output won't be displayed until Enter is pressed, and this was, indeed, my experience. Since switching back and forth between windows, pressing Enter in the console after having sent each command from Vim was killing any joy of using this plugin, I stuck with qtconsole, which refreshes automatically. 

- Modify %PATH% environment variable to enable Anaconda to find gVim:
-- start an Anaconda console
-- Locate where the current environment is stored:
`echo %CONDA_PREFIX%`
-- For (base) environment, it is `C:\ProgramData\Anaconda3`
-- `cd %CONDA_PREFIX%`
-- `mkdir .\etc\conda\activate.d`
-- `mkdir .\etc\conda\deactivate.d`
-- create and edit `.\etc\conda\activate.d\env_vars.bat`:
`@set PATH_BACKUP=%PATH%`
`@set PATH=%PATH%;"C:\Program Files (x86)\Vim\vim81"`
-- create and edit `.\etc\conda\deactivate.d\env_vars.bat`:
`@set PATH=%PATH_BACKUP%`
- Plugin usage:
-- Start an Anaconda console:
-- Start gVim _from the console_, rather than from Start Menu or file context menu (right-click)
-- From the same Anaconda console, execute `jupyter qtconsole`
-- Open a Python file in gVim: `:e c:\path\to\file`
-- Connect Vim to the console: `:JupyterConnect`
  This command is _only_ available if a Python file is open; otherwise, gVim complains about an unknown command.
-- gVim responds "kernel connections successful! pid = NNNNN", but the console does not acknowledge the connection.
-- test the connection by executing the line of code under the cursor with `<localleader>E` (`,E` in my case) or `:JupyterSendRange`
-- Note that it is not necessary to include command jupyter-vim mappings in .vimrc manually: they will be available from the plugin, when everything in set up correctly, and a Python file is open in gVim.

**Comment by wmvanvliet** [comment]
Thanks for the info @willamgld! So in the end, it's not so much that gVim and jupyter need to start from the same console, but that they should run with the same permissions.

**Comment by willamgld** [comment]
I am not sure whether it has to do with permissions or something else... I have not able to get `:JupyterConnect` to succeed when starting gVim either from Start Menu, regular Windows command prompt or from the file context menu (right-click). In those cases, the command fails with the OP error message, after flashing a couple of console windows for a fraction of a second. If your keyboard ninja skills are at the level of catching bullets, as they well should be after using Vim, you would be able to snap one of them with a quick PrintScreen and see this:

![image](https://user-images.githubusercontent.com/53537478/65629098-c555b280-dfa0-11e9-8cd9-cdb22311a4ee.png)

I tried researching into the cause of this error message, and it appears to be a relatively common error not specific to this plugin, but the proposed solutions mostly distill to uninstalling and reinstalling the zmq package in Python, which wipes out approximately 160 of 250 packages in Anaconda and does not restore them automatically, leaving a crippled environment, so I gave up on this path. It looks like something somewhere expects me be building my Python from source code, which I am not going to do. 

The only way I have been able to get everything to work was by starting both gVim and jupyter qtconsole from an Anaconda Prompt. I just realized that it does not have to be the same Prompt instance. I am not sure exactly what makes Anaconda special. I have a vanilla installation with a single base environment, and I have spent way too much time on getting this far to dive into Anaconda. If anyone finds a solution for getting context menu-launched gVim to work with this plugin, I would be happy to learn what it is.

P.S. The best way to start Jupyter qtconsole without tying up the Anaconda prompt is by prepending it with a `START /B`, which is the closest thing to Linux's fork, so I made myself a little handy batch file somewhere in the %PATH% with these commands:
`START /B gvim`
`START /B jupyter qtconsole`
because no self-respecting Vim user has the time to type it every time :) .
